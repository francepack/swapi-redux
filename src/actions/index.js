export const pickRandomCrawl = (films) => ({
  type: 'PICK_CRAWL',
  films
})

export const addFavorite = (id) => ({
  type: 'ADD_FAVORITE',
  id
})

export const deleteFavorite = (id) => ({
  type: 'DELETE_FAVORITE',
  id
})

export const setPeople = (people) => ({
  type: 'SET_PEOPLE',
  people
})

export const setPlanets = (planets) => ({
  type: 'SET_PLANETS',
  planets
})

export const setVehicles = (vehicles) => ({
  type: 'SET_VEHICLES',
  vehicles
})

export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})