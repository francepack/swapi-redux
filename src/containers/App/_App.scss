*,
*:after,
*:before {
  box-sizing: border-box;
}

.App {
  text-align: center;
}

body {
  background-color: $greyBlue;
  margin: 0;
  padding: 0;
  font-family: 'Changa', sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}


a {
  text-decoration: none;
}
