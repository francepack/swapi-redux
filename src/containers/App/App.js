import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Route } from 'react-router-dom';
import PropTypes from 'prop-types';

import { constructPeople, constructPlanets, constructVehicles } from '../../api/helpers';
import { pickRandomCrawl, setPeople, setPlanets, setVehicles } from '../../actions';
import { makeFetch } from '../../api/api';
import Header from '../Header/Header';
import Quote from '../Quote/Quote';
import Control from '../Control/Control';
import CardBox from '../CardBox/CardBox';

export class App extends Component {
  constructor() {
    super();
    this.state = {
      error: ''
    };
  }

  componentDidMount = async () => {
    this.fetchQuote();
    this.fetchPeople();
    this.fetchPlanets();
    this.fetchVehicles();
  }

  fetchQuote = async () => {
    const filmUrl = 'https://swapi.co/api/films/';
    try {
      const films = await makeFetch(filmUrl);
      await this.props.pickRandomCrawl(films.results)
    } catch(error) {
      this.setState({ error: error.message });
    }
  }

  fetchPeople = async () => {
    const peopleUrl = 'https://swapi.co/api/people'
    try {
      const peopleData = await makeFetch(peopleUrl)
      const people =  await constructPeople(peopleData);
      await this.props.setPeople(people)
    } catch(error) {
      this.setState({ error: error.message });
    }
  }

  fetchPlanets = async () => {
    const planetsUrl = 'https://swapi.co/api/planets'
    try {
      const planetData = await makeFetch(planetsUrl)
      const planets =  await constructPlanets(planetData);
      await this.props.setPlanets(planets)
    } catch(error) {
      this.setState({ error: error.message });
    }
  }

  fetchVehicles = async () => {
    const vehiclesUrl = 'https://swapi.co/api/vehicles'
    try {
      const vehicleData = await makeFetch(vehiclesUrl)
      const vehicles =  await constructVehicles(vehicleData);
      await this.props.setVehicles(vehicles)
    } catch(error) {
      this.setState({ error: error.message });
    }
  }

  render() {
    const { quote, title, date } = this.props.quote;
    const quoteContent = { quote, title, date }
    return (
      <div className='App'>
        <Header />
        <Quote {...quoteContent}/>
        <Control />
        {this.state.error && this.state.error}
        {
          <Route exact path='/' render={() => {
            return <CardBox id='home' />
            }}  />
        }
        {
          <Route exact path='/favorites' render={() => {
            return <CardBox id='favorites' />
            }}  />
        }
        {
          <Route exact path='/people' render={() => {
            return <CardBox id='people' />
            }}  />
        }
        {
          <Route exact path='/planets' render={() => {
            return <CardBox id='planets' />
            }}  />
        }
        {
          <Route exact path='/vehicles' render={() => {
            return <CardBox id='vehicles' />
            }}  />
        }

      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  quote: state.quote
})

export const mapDispatchToProps = (dispatch) => ({
  pickRandomCrawl: (films) => dispatch(pickRandomCrawl(films)),
  setPeople: (people) => dispatch(setPeople(people)),
  setPlanets: (planets) => dispatch(setPlanets(planets)),
  setVehicles: (vehicles) => dispatch(setVehicles(vehicles))
})

export default connect(mapStateToProps, mapDispatchToProps)(App);
