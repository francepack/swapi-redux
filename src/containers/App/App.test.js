import React from 'react';
import { shallow } from 'enzyme';
import { setPeople, setPlanets, setVehicles } from '../../actions'
import ReactDOM from 'react-dom';
import { App, mapStateToProps, mapDispatchToProps} from './App';

describe('App', () => {
  let wrapper;
  let mockProps = { 
    quote: { quote: 'In a galaxy closeby...', title: 'Star Wars 16, Newest Hopes', date: 'December 2030' }
  }
  beforeEach(() => {
    wrapper = shallow(
      <App {...mockProps} />
    )
  });  
  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot()
  });

  describe('mapStateToProps', () => {
    it('should return an object with a quote', () => {
      const mockState = {
        people: [],
        planets: [],
        vehicles: [],
        quote: {quote: 'In a galaxy pretty closeby...'}
      };

      const expectedState = {
        quote: {quote: 'In a galaxy pretty closeby...'}
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expectedState);
    });
  });

  describe('mapDispatchToProps', () => {

    it('should call dispatch with a setPeople action', () => {
      const mockPeople = [{name: 'Mason'}, {name: 'Luke'}];
      const dispatch = jest.fn();
      const actionToDispatch = setPeople(mockPeople);
      const mappedProps = mapDispatchToProps(dispatch);
      mappedProps.setPeople(mockPeople);
      expect(dispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a setPlanets action', () => {
      const mockPlanets = [{name: 'Mars'}, {name: 'Hoth'}];
      const dispatch = jest.fn();
      const actionToDispatch = setPlanets(mockPlanets);
      const mappedProps = mapDispatchToProps(dispatch);
      mappedProps.setPlanets(mockPlanets);
      expect(dispatch).toHaveBeenCalledWith(actionToDispatch);
    });
    
    it('should call dispatch with a setVehicles action', () => {
      const mockVehicles = [{name: 'tie-fighter'}, {name: 'x-wing'}];
      const dispatch = jest.fn();
      const actionToDispatch = setVehicles(mockVehicles);
      const mappedProps = mapDispatchToProps(dispatch);
      mappedProps.setVehicles(mockVehicles);
      expect(dispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

});
