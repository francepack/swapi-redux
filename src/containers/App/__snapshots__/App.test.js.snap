// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    quote={
      Object {
        "date": "December 2030",
        "quote": "In a galaxy closeby...",
        "title": "Star Wars 16, Newest Hopes",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header />,
        <Quote
          date="December 2030"
          quote="In a galaxy closeby..."
          title="Star Wars 16, Newest Hopes"
        />,
        <Control />,
        "",
        <Route
          exact={true}
          path="/"
          render={[Function]}
        />,
        <Route
          exact={true}
          path="/favorites"
          render={[Function]}
        />,
        <Route
          exact={true}
          path="/people"
          render={[Function]}
        />,
        <Route
          exact={true}
          path="/planets"
          render={[Function]}
        />,
        <Route
          exact={true}
          path="/vehicles"
          render={[Function]}
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "date": "December 2030",
          "quote": "In a galaxy closeby...",
          "title": "Star Wars 16, Newest Hopes",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/favorites",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/people",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/planets",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/vehicles",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header />,
          <Quote
            date="December 2030"
            quote="In a galaxy closeby..."
            title="Star Wars 16, Newest Hopes"
          />,
          <Control />,
          "",
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />,
          <Route
            exact={true}
            path="/favorites"
            render={[Function]}
          />,
          <Route
            exact={true}
            path="/people"
            render={[Function]}
          />,
          <Route
            exact={true}
            path="/planets"
            render={[Function]}
          />,
          <Route
            exact={true}
            path="/vehicles"
            render={[Function]}
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "date": "December 2030",
            "quote": "In a galaxy closeby...",
            "title": "Star Wars 16, Newest Hopes",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/favorites",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/people",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/planets",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/vehicles",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
